protocol/testing/containertest/testnet_test.go

```diff
@@ -291,7 +290,54 @@ func TestUpgrade(t *testing.T) {
 	require.NoError(t, err)
 	defer testnet.MustCleanUp()
 	node := testnet.Nodes["alice"]
+	nodeAddress := constants.AliceAccAddress.String()
+	err = upgradeTestnet(nodeAddress, t, node)
+	require.NoError(t, err)
+}
+
+func TestStateUpgradeForPerpetualMarketType(t *testing.T) {
+	testnet, err := NewTestnetWithPreupgradeGenesis()
+	require.NoError(t, err, "failed to create testnet - is docker daemon running?")
+	err = testnet.Start()
+	require.NoError(t, err)
+	defer testnet.MustCleanUp()
+	node := testnet.Nodes["alice"]
+
+	nodeAddress := constants.AliceAccAddress.String()
+	queryClient := perpetuals.NewQueryClient
+	requestFunction := perpetuals.QueryClient.Perpetual
+	request := &perpetuals.QueryPerpetualRequest{
+		Id: 1,
+	}
+
+	resp, err := Query(
+		node,
+		queryClient,
+		requestFunction,
+		request,
+	)
+	require.NoError(t, err)
+	perpetualObject := &perpetuals.QueryPerpetualResponse{}
+	err = proto.UnmarshalText(resp.String(), perpetualObject)
+	require.NoError(t, err)
+	assert.Equal(t, perpetuals.PerpetualMarketType_PERPETUAL_MARKET_TYPE_UNSPECIFIED, perpetualObject.Perpetual.Params.MarketType)
+
+	err = upgradeTestnet(nodeAddress, t, node)
+	require.NoError(t, err)
 
+	resp, err = Query(
```

Comment:
Might be useful to put everything below in a separate helper function `postUpgradeChecks`